File: server/src/services/blockchainService.ts
import Web3 from 'web3';
import { AbiItem } from 'web3-utils';

const web3 = new Web3(process.env.ETHEREUM_NODE_URL || 'http://127.0.0.1:7545');
const contractABI: AbiItem[] = [
    {
        type: 'function',
        name: 'setRecord',
        stateMutability: 'nonpayable',
        inputs: [{ name: '_recordHash', type: 'string' }],
        outputs: []
    },
    {
        type: 'function',
        name: 'getRecordOwner',
        stateMutability: 'view',
        inputs: [{ name: '_recordHash', type: 'string' }],
        outputs: [{ name: '', type: 'address' }]
    },
    {
        type: 'function',
        name: 'verifyRecord',
        stateMutability: 'view',
        inputs: [
            { name: '_recordHash', type: 'string' },
            { name: '_owner', type: 'address' }
        ],
        outputs: [{ name: '', type: 'bool' }]
    }
];
const contractAddress = process.env.CONTRACT_ADDRESS || ''; // Add your contract address

const getContract = () => {
    if (!contractAddress) {
        throw new Error("CONTRACT_ADDRESS environment variable not set.");
    }
    return new web3.eth.Contract(contractABI, contractAddress);
};

export const blockchainService = {
    storeRecordHash: async (hash: string): Promise<{ transactionHash: string, blockNumber: number }> => {
        const contract = getContract();
        const accounts = await web3.eth.getAccounts();
        const gasEstimate = await contract.methods.setRecord(hash).estimateGas({ from: accounts[0] });
        const gas = Number(gasEstimate);
        
        const receipt = await contract.methods.setRecord(hash).send({ from: accounts[0], gas: gas.toString() });
        const blockNumber = typeof receipt.blockNumber === 'bigint'
            ? Number(receipt.blockNumber)
            : Number(receipt.blockNumber ?? 0);

        return {
            transactionHash: receipt.transactionHash,
            blockNumber
        };
    },

    verifyRecordHash: async (hash: string, transactionHash: string): Promise<boolean> => {
        const contract = getContract();
        const accounts = await web3.eth.getAccounts();
        
        // Fetch the record from the blockchain using the hash as a key
        const storedHashOwner = String(await contract.methods.getRecordOwner(hash).call({ from: accounts[0] }));

        if (storedHashOwner.toLowerCase() === '0x0000000000000000000000000000000000000000') {
            return false; // Hash not found on the blockchain
        }

        // Additionally, you can verify against the transaction if needed, but getting the owner is a direct check.
        const tx = await web3.eth.getTransaction(transactionHash);
        if (!tx?.input) {
            return false;
        }
        const inputData = contract.methods.setRecord(hash).encodeABI().toLowerCase();
        
        // This check is a bit complex and might not be fully reliable if other data is in the transaction.
        // The primary verification should be fetching the record by its hash.
        return tx.input.toLowerCase() === inputData;
    }
};

File: python_server/app/services/blockchain_service.py
from web3 import Web3
import os
import json
from dotenv import load_dotenv

load_dotenv()

NODE_URL = os.getenv("ETHEREUM_NODE_URL", "http://127.0.0.1:7545")
CONTRACT_ADDRESS = os.getenv("CONTRACT_ADDRESS")

w3 = Web3(Web3.HTTPProvider(NODE_URL))

ABI = [
    {
        "type": "function",
        "name": "setRecord",
        "stateMutability": "nonpayable",
        "inputs": [{"name": "_recordHash", "type": "string"}],
        "outputs": []
    },
    {
        "type": "function",
        "name": "getRecordOwner",
        "stateMutability": "view",
        "inputs": [{"name": "_recordHash", "type": "string"}],
        "outputs": [{"name": "", "type": "address"}]
    },
    {
        "type": "function",
        "name": "verifyRecord",
        "stateMutability": "view",
        "inputs": [
            {"name": "_recordHash", "type": "string"},
            {"name": "_owner", "type": "address"}
        ],
        "outputs": [{"name": "", "type": "bool"}]
    }
]

def get_contract():
    if not CONTRACT_ADDRESS:
        raise ValueError("CONTRACT_ADDRESS not set in .env")
    return w3.eth.contract(address=CONTRACT_ADDRESS, abi=ABI)

async def store_record_hash(hash_str: str):
    if not w3.is_connected():
        raise Exception("Blockchain node not connected")
    
    contract = get_contract()
    accounts = w3.eth.accounts
    if not accounts:
        raise Exception("No accounts available in node")
    
    # In Ganache, we can usually just send from accounts[0]
    tx_hash = contract.functions.setRecord(hash_str).transact({'from': accounts[0]})
    receipt = w3.eth.wait_for_transaction_receipt(tx_hash)
    
    return {
        "transactionHash": receipt.transactionHash.hex(),
        "blockNumber": receipt.blockNumber
    }

async def verify_record_hash(hash_str: str, tx_hash_str: str):
    contract = get_contract()
    accounts = w3.eth.accounts
    
    # 1. Check owner by hash
    owner = contract.functions.getRecordOwner(hash_str).call()
    if owner == '0x0000000000000000000000000000000000000000':
        return False
        
    # 2. Check transaction input
    tx = w3.eth.get_transaction(tx_hash_str)
    if not tx:
        return False
        
    # Simplified check: just verifying the transaction exists and the hash is owned
    return True
